[Zachary Chan] iP

Ardentsoul opened the pull request:
## **Sussus Dukus/Duke**
> "Manage your tasks so that you don't seem sus." - Red Imposter ([amongus](https://tenor.com/view/among-us-twerk-yellow-ass-thang-gif-18983570))

Duke makes sure that you don't get called out being a sussy baka for not doing your tasks. It's,
- definitely not an imposter
- easy to learn for crewmates
- ~~FAST~~ ***I AM SPEED*** level of usage

All you gotta do is,

1. download it from this definitely not sus [website](https://github.com/Ardentsoul/ip/releases).
2. click on the latest version (currently v0.11) and download the jar file.
3. run cmd.
4. add, delete and find your tasks.
5. let this **CHAD** bot manage your tasks for you 🤤💦 

Best part is this is **FREE**! 💯 

Features:

- [x] Manage your sussy tasks
- [x] Manage your todos, deadlines and events
- [ ] Reminders (coming soon maybe?)

#### If you program in the sussy language known as Java, you can use it to practice Java too. Here's the `main` method:
```java
public class Duke {
    public static void main(String[] args) {
        String home = System.getProperty("user.home");
        new Duke(home,"/DukeSaveDirectory/DukeSaveFile.txt").run();
    }
}
```

Commit:
Add Gradle support

Ardentsoul committed:
Added a greet, echo, exit function.

Ardentsoul committed:
Added an add and list function
implemented an arraylist to store the task & created a Task class

Ardentsoul committed:
Added mark and unmark functions
Format changes to Duke.java
Added support for checking completion of tasks

Ardentsoul committed:
Added Deadline, Event and Todo classes

Ardentsoul committed:
Added todo, deadline, event function

Ardentsoul committed:
Removed mascot

Ardentsoul committed:
Testing the Automated Text UI Testing

Ardentsoul committed:
Updated Text-ui-test

Ardentsoul committed:
Added 4 Exception classes
Updated Duke.java to catch errors

Ardentsoul committed:
test cases for Level-5

Ardentsoul committed:
Created several methods for encapsulation
Added a delete function

Ardentsoul committed:
test cases for level-6

Ardentsoul committed:
Encapsulated the list() function

Ardentsoul committed:
Renamed list() to printList()
Grammar changes

Ardentsoul committed:
Handled 1 more exception & text changes

Deadline and Event function catches error when there is no space after by/at.
Changed some String outputs to suit among us theme.

Ardentsoul committed:
Add Javadocs to Duke.java

There were no descriptions to the methods in Duke.java

Adding the Javadocs allows us to better understand what the methods
are meant to do.

Ardentsoul committed:
Complete Level-7

Added a auto save function

Ardentsoul committed:
Update Level-7

Changed method name from readTxtFile() to readTextFile()

Ardentsoul committed:
Update Level-7

Edited message in error handling for txt not found

Ardentsoul committed:
Update Level-7

Updated the error messages if the directory or txt is not found

Ardentsoul committed:
Complete Level-8

A formatDateTime method was added to take a String input, parse it,
and output the date in a readable format. This is applicable to
deadline and event commands.

Ardentsoul committed:
Merge branch 'branch-Level-7'

Ardentsoul committed:
Resolve merging branch-Level-8 to master

Kept the methods implemented in branch-Level-8 and master

Ardentsoul committed:
Add temporary classes

This is just for saving a copy

Ardentsoul committed:
Update partially correct Duke and Fix several bugs

Currently updating OOP and included creation of directory and file
if they do not exist. Also made it OS-independent.

Ardentsoul committed:
Update A-MoreOOP

Restructured the original code to be more OOP. Fixed Level-7 &
Level-8 at the same time.

Ardentsoul committed:
Update A-MoreOOP

The directory and txt file names have been changed to be
more specific instead of '/data/tasks.txt'.

Ardentsoul committed:
Update A-Packages

Divided the classes in to 5 packages.

Ardentsoul committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Ardentsoul committed:
Update Level-8 and A-MoreOOP

Deadline and Event classes take in LocalDate and Local Time.
Restructured the read and write of the txt file.

Ardentsoul committed:
Add A-JUnit tests for Parser and Task

Applied Gradle to test own testcases.

Checked for String conversion of all 3 types of tasks, isExit(),
and exception thrown when input date/time is wrong

Ardentsoul committed:
Implement a find feature to Duke

A find function has been added. This is will search for words
that contain the keyword in each Task.

Ardentsoul committed:
Update A-JavaDoc partially

Half of the JavaDoc is complete.

Ardentsoul committed:
Update A-JavaDoc fully

Added all JavaDocs to all methods currently implemented.

Ardentsoul committed:
Update A-CodingStandard

Changed some code to fit the coding standard.

Ardentsoul committed:
Merge branch 'A-JavaDoc'

Ardentsoul committed:
Merge branch 'A-CodingStandard'

Ardentsoul committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/command/Command.java

Ardentsoul committed:
Update Javadocs for Find feature

Added in JavaDocs for find method and prints a message when
nothing is keyed after "find".
Edited a few comments in Parser.java.

Ardentsoul committed:
Update A-CodingStandard

Updated coding standard for imports in Parser class.

Ardentsoul committed:
Update A-CodingStandard

Changed Parse to Parses in header comments for parse method.

Ardentsoul committed:
Update A-CodingStandards

Encapsulated the logic of mark, unmark, todo, deadline and event
into methods.

Ardentsoul committed:
Update A-Jar

Update A-Jar so that it can execute in cmd.

Add manifest file.

Ardentsoul committed:
Rectify A-Jar & A-JUnit

Shadow Gradle was rectified to the correct relative path for Duke
and updated the JUnit test files.

Ardentsoul committed:
Update A-CheckStyle halfway

Updated checkstyle in code. It is still partially done.

Ardentsoul committed:
Update A-Checkstyle completely

Finished correcting the checkstyle for the code.

Ardentsoul committed:
Add Level-10 GUI for Duke

Added GUI using FXML and checked checkstyle.

Ardentsoul committed:
Merge branch 'branch-Level-10'

Ardentsoul committed:
Update Level-10 GUI

Made the GUI look slightly better and changed the image of the
user.

lamwj98 commented on src/main/java/duke/command/PrintCommand.java:
```diff
@@ -0,0 +1,27 @@
+     */
+    @Override
+    public String execute(TaskList taskList) {
+        String firstMessage = "Here are the tasks in your device:";
+        return firstMessage + "\n   " + taskList.toString();
```
Would it be better to just return concat all the required messages together and return the variable? Is there a need to split it into two portion ?

lamwj98 commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,35 @@
+    public String execute(TaskList taskList) {
+        try {
+            taskList.unmark(num);
+            String firstMessage = "Surely you aren't the imposter... right??";
+            String secondMessage = taskList.toString();
+            return firstMessage + "\n   " + secondMessage;
```
Same as previous, would it be better to put concat under one variable and return that variable?


```suggestion
            String message = "Surely you aren't the imposter... right??" + "\n" + taskList.toString();
            return message;
```

lamwj98 commented on src/main/java/duke/operations/Parser.java:
```diff
@@ -0,0 +1,197 @@
+            throw new EmptyInputException();
+        } else if (input.equalsIgnoreCase("bye")) {
+            return new ExitCommand();
+        } else if (input.equalsIgnoreCase("list")) {
+            return new PrintCommand();
+        } else if (firstWord.equalsIgnoreCase("mark")) {
+            return handleMarkInput(strs[1]);
+        } else if (firstWord.equalsIgnoreCase("unmark")) {
+            return handleUnmarkInput(strs[1]);
+        } else if (firstWord.equalsIgnoreCase("todo")) {
+            String subString = input.substring(4).trim(); // take the remaining of the input String after "todo"
+            if (subString.length() == 0) { // Check for empty description
+                throw new IncompleteInputException(firstWord);
+            } else {
+                return handleToDoInput(subString);
+            }
+        } else if (firstWord.equalsIgnoreCase("deadline")) {
+            String subString = input.substring(8).trim(); // take the remaining of the input String after "deadline"
+            if (subString.length() == 0) { // Check for empty description
+                throw new IncompleteInputException(firstWord);
+            } else {
+                return handleDeadlineInput(subString);
+            }
+        } else if (firstWord.equalsIgnoreCase("event")) {
+            String subString = input.substring(5).trim(); // take the remaining of the input String after "event"
+            if (subString.length() == 0) { // Check for empty description
+                throw new IncompleteInputException(firstWord);
+            } else {
+                return handleEventInput(subString);
+            }
+        } else if (firstWord.equalsIgnoreCase("delete")) {
+            int listIndex = Integer.parseInt(strs[1]); // retrieve the index after delete
+            Task taskToBeDeleted = TaskList.TASK_ARRAY_LIST.get(listIndex - 1);
+            return new DeleteCommand(taskToBeDeleted);
+        } else if (firstWord.equalsIgnoreCase("find")) {
+            String keyword = input.substring(4).trim(); // take the remaining of the input String
+            return new FindCommand(keyword);
+        } else {
+            throw new UnknownInputException();
+        }
```
would it be better to use a switch statement?

lamwj98 reviewed (commented):
Overall, good is easy to read with great naming. Just think that it would have been better to concat the strings together and return instead of splitting them into multiple variables and then concat them at point of return.

Ardentsoul committed:
Update build.gradle

The mainClassName under application was changed to
duke.main.Launcher so that the jar file can be run.

Ardentsoul committed:
Remove Emojis from Exception classes

Emojis were removed since they were not being able to be displayed
on the GUI.

Ardentsoul committed:
Refactor methods in Parser, Storage, and TaskList classes

Parser and Storage classes have long methods, uses magic numbers,
and does not apply Single Level of Abstraction Principle (SLAP)
properly. TaskList class contains dead code.

The common behaviours across Parser and Storage class cause some
code duplication and does not maximize readability.

Extracting the common behaviour between Parser and Storage classes
and splitting them into several short methods allows us to minimize
code duplication and improve readability. For example, the magic
numbers in these two classes can be avoided during the the handling
of the three different task events.

Let's create several new methods to contain the logic of handling
these three tasks(todo, deadline, event) called handleTodo,
handleDeadline, handleEvent.

SLAP is preferable in this situation because it makes the parse
method more readable.

TeddYE reviewed (commented):
Overall, I found your code easy to read.

TeddYE commented on src/main/java/duke/operations/Parser.java:
```diff
@@ -0,0 +1,197 @@
+
+        if (input.equals("")) {
+            throw new EmptyInputException();
+        } else if (input.equalsIgnoreCase("bye")) {
```
Would using switch case be more suitable in this scenario?

TeddYE commented on src/main/java/duke/operations/Ui.java:
```diff
@@ -0,0 +1,41 @@
+     * @return the welcome message for Duke.
+     */
+    public static String showWelcome() {
+        String getWelcomeMsgOne = WELCOME_MESSAGE_ONE;
```
Why did you split your welcome message into 2 different strings?

TeddYE commented on src/main/java/duke/operations/Ui.java:
```diff
@@ -0,0 +1,41 @@
+    public static String showWelcome() {
+        String getWelcomeMsgOne = WELCOME_MESSAGE_ONE;
+        String getWelcomeMsgTwo = WELCOME_MESSAGE_TWO;
+        return getWelcomeMsgOne + "\n" + getWelcomeMsgTwo;
```
Can't you just return WELCOME_MESSAGE_ONE + "\n" + WELCOME_MESSAGE_TWO? Why create two local vars?

TeddYE commented on src/main/java/duke/operations/TaskList.java:
```diff
@@ -0,0 +1,129 @@
+     * @param num total number of tasks.
+     * @return the String to be printed.
+     */
+    public String totalTasks(int num) {
```
From what i notice, ur num in simply the size of the arraylist which u can get but calling getTaskArrayListSize() or TASK_ARRAY_LIST.size(). Why did you input it as an argument?

Ardentsoul committed:
Refactor parse method in Parser

Parse method is not as reader friendly.
Ui class contains duplicate code.

The common behaviours across Parser and Ui class do not maximize
readability.

Extracting the common behaviour between Parser and Ui classes
and converting them into switch case statements allows us to
improve readability. For example, using switch case statements to
handle the logic of the parse method makes the code more readable.

Let's refactor the parse method to be a switch case statement
instead of if-else statements and delete duplicate code in Ui
class.

Switch case statements are preferable in this situation because
it makes the mode more readable.

Ardentsoul committed:
Update Mark and Unmark Command classes

Concatenated the Strings under one variable and return that
variable.

Ardentsoul committed:
Update Mark and Unmark Command classes

Concatenated the Strings under one variable and return that
variable.

Ardentsoul committed:
Add assert method into delete method

Added an assert feature to check whether index keyed after delete
is more than 0.

Ardentsoul committed:
Merge pull request #2 from Ardentsoul/branch-A-CodeQuality

Branch a code quality

Ardentsoul committed:
Merge branch 'master' into branch-A-Assertions

Ardentsoul committed:
Merge pull request #1 from Ardentsoul/branch-A-Assertions

Add assert method into delete method

Ardentsoul committed:
Add Exit command for GUI

Exit command has a behaviour (e.g. hasExit).

Closing the application without pressing
the 'X' button on the top left of the GUI allows the user
an alternative to close the application.

Implementing this behaviour gives the user an alternative to exit
the application with the Exit command. Furthermore, a delay of
1.5 seconds has been added after the exit command is typed by the
user. This means that after 1.5 seconds from entering the
Exit command, the application will automatically close.

Ardentsoul committed:
Delete Text-Ui-test folder

This folder is not needed anymore since Duke is UI based now.

Ardentsoul committed:
Add Tag and Untag features

Additional features such as tagging and untagging have been added
into Duke bot. This is so as to make the bot more versatile and
handle more user inputs.

These features give more flexibility to the bot and
users will also have additional functions to use when trying the
bot.

Implementing these features allow the user to be able to add
additional tags to their tasks to further express his/her
short comments or untag their comment. For example,
tagging/untagging can be done for all tasks that are present in
the user's list.

Let's create handleTag and handleUntag methods to process the logic
of these features in the Parser class.

Using composition is preferable over inheritance when adding the
Tag class into the Task class because a Task HAS A Tag. Thus,
Tag is an attribute of Task.

Ardentsoul committed:
Merge branch 'branch-A-Extension-Tagging'

Ardentsoul committed:
Add Javadoc comment for Tag class

Added missing javadoc comment for Tag class.

Ardentsoul committed:
Add missing Javadoc

Added missing @param javadoc for IncompleteInputException class.

Ardentsoul committed:
Update GUI

Updated GUI to be more aesthetically pleasing to look at. GUI has
been tweaked to imitate dark mode.

Ardentsoul committed:
Add A-UserGuide

Added a user guide so that users will know how to use the program.

Ardentsoul committed:
Set theme jekyll-theme-midnight

Ardentsoul committed:
Merge branch 'master' of https://github.com/Ardentsoul/ip

Ardentsoul committed:
Rename Ui file

Ardentsoul committed:
Rename Ui file

Removed .PNG extension.

Ardentsoul committed:
Update README.txt

Updated README.txt to follow github description.
